Paintroid/src/main/java/org/catrobat/paintroid/command/implementation/CommandManagerImplementation.java

```diff
@@ -19,147 +19,513 @@
 
 package org.catrobat.paintroid.command.implementation;
 
+import android.util.Pair;
+
+import org.catrobat.paintroid.command.Command;
+import org.catrobat.paintroid.command.CommandManager;
+import org.catrobat.paintroid.command.LayerBitmapCommand;
+import org.catrobat.paintroid.eventlistener.ChangeActiveLayerEventListener;
+import org.catrobat.paintroid.eventlistener.LayerEventListener;
+import org.catrobat.paintroid.eventlistener.RefreshLayerDialogEventListener;
+import org.catrobat.paintroid.eventlistener.UpdateTopBarEventListener;
+import org.catrobat.paintroid.tools.Layer;
+
+import java.util.ArrayList;
 import java.util.LinkedList;
+import java.util.ListIterator;
 import java.util.Observable;
 import java.util.Observer;
 
-import org.catrobat.paintroid.PaintroidApplication;
-import org.catrobat.paintroid.command.Command;
-import org.catrobat.paintroid.command.CommandManager;
-import org.catrobat.paintroid.command.UndoRedoManager;
-import org.catrobat.paintroid.command.UndoRedoManager.StatusMode;
-import org.catrobat.paintroid.dialog.IndeterminateProgressDialog;
+public class CommandManagerImplementation implements CommandManager, Observer
+{
+	private static final int INIT_APP_lAYER_COUNT = 1;
+
+	enum CommandType {COMMIT_LAYER_BITMAP_COMMAND
+		,ADD_LAYER
+		,REMOVE_LAYER
+		,MERGE_LAYERS
+		,CHANGE_LAYER_VISIBILITY
+		,LOCK_LAYER
+		,RENAME_LAYER}
+
 
-import android.graphics.Bitmap;
-import android.graphics.Bitmap.Config;
+	enum Action {UNDO, REDO}
 
-public class CommandManagerImplementation implements CommandManager, Observer {
-	private static final int MAX_COMMANDS = 512;
+	private LinkedList<Pair<CommandType, LayerCommand>> mLayerCommandList;
+	private LinkedList<Pair<CommandType, LayerCommand>> mLayerUndoCommandList;
+	private ArrayList<LayerBitmapCommand> mLayerBitmapCommands;
```

Comment:
what's the difference between mLayerBitmapCommands and mLayerCommandList?
There should be a clear difference in the name to identify the use of both lists if needed